src/common/hashtable.c

```diff
@@ -0,0 +1,351 @@
+*****************************************/
+onf_ht_hash_table* onf_ht_new_sized(const int base_size) {
+    onf_ht_hash_table* ht = malloc(sizeof(onf_ht_hash_table));
+    if(NULL == ht)
+    {
+        return NULL;
+    }
+
+    ht->base_size = base_size;
+    ht->size = onf_next_prime(ht->base_size);
+
+    ht->count = 0;
+    ht->items = calloc((size_t)ht->size, sizeof(onf_ht_item*));
+    if(NULL == ht->items)
+    {
+        free(ht);
+	return NULL;
+    }
+
+    return ht;
+}
+
+/*****************************************
+Name:  onf_ht_resize
+brief: resize hash table with new size
+return: 
+*****************************************/
+int onf_ht_resize(
+         onf_ht_hash_table* ht, const int base_size) 
+{
+    if (base_size < ONF_HT_INITIAL_BASE_SIZE) {
+        return E_FAIL;
+    }
+
+    onf_ht_hash_table* new_ht = onf_ht_new_sized(base_size);
+    if(NULL == new_ht)
+    {
+        return E_ALLOC_FAILED;
+    }
+
+    for (int i = 0; i < ht->size; i++) {
+        onf_ht_item* item = ht->items[i];
+        if (item != NULL && item != &ONF_HT_DELETED_ITEM) {
+            if(SUCCESS != onf_ht_insert(new_ht, item->key_len,
+	      item->val_len, item->key, item->value))
+	    {
+	        return E_FAIL;
+	    }
+        }
+    }
+
+    ht->base_size = new_ht->base_size;
+    ht->count = new_ht->count;
+
+    // To delete new_ht, we give it ht's size and items 
+    const int tmp_size = ht->size;
+    ht->size = new_ht->size;
+    new_ht->size = tmp_size;
+
+    onf_ht_item** tmp_items = ht->items;
+    ht->items = new_ht->items;
+    new_ht->items = tmp_items;
+
+    onf_ht_del_hash_table(new_ht);
+    return SUCCESS;
+}
+
+/*****************************************
+Name:  onf_ht_resize_up
+brief: resize up 
+return: 
+*****************************************/
+int onf_ht_resize_up(onf_ht_hash_table* ht) {
+    const int new_size = ht->base_size * 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_resize_down
+brief: resize down
+return: 
+*****************************************/
+int onf_ht_resize_down(onf_ht_hash_table* ht) {
+    const int new_size = ht->base_size / 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_del_item
+brief: delete hash item
+return: 
+*****************************************/
+void onf_ht_del_item(onf_ht_item* i) {
+    free(i->key);
+    free(i->value);
+    free(i);
+}
+
+/*****************************************
+Name:  onf_ht_del_hash_table
+brief: delete hash table
+return: 
+*****************************************/
+void onf_ht_del_hash_table(onf_ht_hash_table* ht) {
+    for (int i = 0; i < ht->size; i++) {
+        onf_ht_item* item = ht->items[i];
+        if (item != NULL) {
+            onf_ht_del_item(item);
+        }
+    }
+    free(ht->items);
+    free(ht);
+}
+
+/*****************************************
+Name:  onf_ht_hash
+brief: delete hash table
```

Comment:
Can you update brief. ? Looks copy/paste error

Reply:
done